### DISCUSSION TITLE
Single-pass ZIP generation

### DISCUSSION THREAD

**Discussion by tanriol** [zip-rs/zip-old#339] [Feature Design]
`ZipWriter` currently expects the underlying stream to be seekable. The seekability is used to update the local file header after the file is complete (and also to determine some offsets).

The ZIP file format specification supports writing an archive in one pass without seeking. This mode may be useful in some cases (for example, streaming a ZIP archive of several large files without generating the archive ahead of time).

Would you be interested in implementing this mode and lifting the seekability requirement?

**Comment by mvdnes** [comment]
I would love to have this, but lack the time and interest unfortunately.

It should not be to hard, as the only thing to do is to write the sizes and CRC after the data instead of before. The only real issue lies in determining the compressed file size, as that also uses the Seek trait at the moment.

**Comment by gyscos** [comment]
When writing the file, we can remember how much data was written to compute the compressed size.
When reading, it's probably fine to require the `Seek` trait as we need to jump between the files and the central directory anyway.

**Comment by tanriol** [comment]
Technically single-pass reading is possible too as most information from the central directory is duplicated in the file headers too, IIRC.

**Comment by Plecra** [comment]
I'm afraid I'm going to have to put this one on the backburner for a while. `ZipWriter` isn't flexible enough to easily implement this and I've already bounced a couple of PRs that added `Seek`-free implementations. We'll return to it when we look at evolving the API again.
